### PR TITLE
Added expression support for type based assertions

### LINKED ISSUE #908: Add expression support for type based assertions
Currently the following assertion methods doesn't have an equivalent expression in YAML/ JSON.

- IsArray
- IsBoolean
- IsDateTime
- IsInteger
- IsNumeric

We should add support for these.

### PR DESCRIPTION
**Pull request by ArmaanMcleod**
## PR Summary

Fixes #908

Added the following expressions:

- `IsArray`
- `IsBoolean`
- `IsDateTime`
- `IsInteger`
- `IsNumeric`

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Code changes**
  - [x] Have unit tests created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Microsoft/PSRule/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

File: tests/PSRule.Tests/SelectorTests.cs [hunk 1]
```diff
@@ -717,6 +717,302 @@ public void IsStringExpression(string type, string path)
...
+            var actual6 = GetObject((name: "value", value: PSObject.AsPSObject(DateTime.Now)));
+            var actual7 = GetObject((name: "value", value: new JValue("2021-04-03T15:00:00.00+10:00")));
+            var actual8 = GetObject((name: "value", value: long.MaxValue));
+            var actual9 = GetObject();
+
+            // Without conversion
+            var isDateTimeTrue = GetSelectorVisitor($"{type}IsDateTimeTrue", GetSource(path), out _);
+            var isDateTimeFalse = GetSelectorVisitor($"{type}IsDateTimeFalse", GetSource(path), out _);
+
+            // isArray: true
```
Comment (BernieWhite) [inline tests/PSRule.Tests/SelectorTests.cs:836]:
```suggestion
            // isDateTime: true
```

File: tests/PSRule.Tests/SelectorTests.cs [hunk 2]
```diff
@@ -717,6 +717,302 @@ public void IsStringExpression(string type, string path)
...
+            Assert.False(isDateTimeTrue.Match(actual2));
+            Assert.False(isDateTimeTrue.Match(actual3));
+            Assert.True(isDateTimeTrue.Match(actual4));
+            Assert.False(isDateTimeTrue.Match(actual5));
+            Assert.True(isDateTimeTrue.Match(actual6));
+            Assert.False(isDateTimeTrue.Match(actual7));
+            Assert.False(isDateTimeTrue.Match(actual8));
+            Assert.False(isDateTimeTrue.Match(actual9));
+
+            // isArray: false
```
Comment (BernieWhite) [inline tests/PSRule.Tests/SelectorTests.cs:847]:
```suggestion
            // isDateTime: false
```

File: tests/PSRule.Tests/SelectorTests.cs [hunk 3]
```diff
@@ -717,6 +717,302 @@ public void IsStringExpression(string type, string path)
...
+            Assert.False(isDateTimeFalse.Match(actual6));
+            Assert.True(isDateTimeFalse.Match(actual7));
+            Assert.True(isDateTimeFalse.Match(actual8));
+            Assert.False(isDateTimeFalse.Match(actual9));
+
+            // With conversion
+            var isDateTimeConvertTrue = GetSelectorVisitor($"{type}IsDateTimeTrueWithConversion", GetSource(path), out _);
+            var isDateTimeConvertFalse = GetSelectorVisitor($"{type}IsDateTimeFalseWithConversion", GetSource(path), out _);
+
+            // isArray: true
```
Comment (BernieWhite) [inline tests/PSRule.Tests/SelectorTests.cs:862]:
```suggestion
            // isDateTime: true
```

File: tests/PSRule.Tests/SelectorTests.cs [hunk 4]
```diff
@@ -717,6 +717,302 @@ public void IsStringExpression(string type, string path)
...
+            Assert.True(isDateTimeConvertTrue.Match(actual2));
+            Assert.True(isDateTimeConvertTrue.Match(actual3));
+            Assert.True(isDateTimeConvertTrue.Match(actual4));
+            Assert.False(isDateTimeConvertTrue.Match(actual5));
+            Assert.True(isDateTimeConvertTrue.Match(actual6));
+            Assert.True(isDateTimeConvertTrue.Match(actual7));
+            Assert.False(isDateTimeConvertTrue.Match(actual8));
+            Assert.False(isDateTimeConvertTrue.Match(actual9));
+
+            // isArray: false
```
Comment (BernieWhite) [inline tests/PSRule.Tests/SelectorTests.cs:873]:
```suggestion
            // isDateTime: false
```

**Review by BernieWhite — APPROVED** [review]
@ArmaanMcleod A few minor suggestions but otherwise all good to merge. Nice work!

**Comment by ArmaanMcleod** [comment]
> @ArmaanMcleod A few minor suggestions but otherwise all good to merge. Nice work!

Thanks @BernieWhite 🎉. Have applied your suggestions, thanks for picking that up 🙂.

**Pull request merged**